### PR TITLE
5256 added user-agent header for remote cache (http)

### PR DESCRIPTION
**Pull request by ishikhman**
* reverted the rollback
* changed user-agent value for development environment to "bazel/dev"

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/analysis/BlazeVersionInfo.java [hunk 1]
```diff
@@ -117,12 +119,12 @@ public String getReleaseName() {
   }
 
   /**
-   * Returns the version, if any, or HEAD. The returned version number is easier to
+   * Returns the version, if any, or an empty String. The returned version number is easier to
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/analysis/BlazeVersionInfo.java]: {@code ""}

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java [hunk 2]
```diff
@@ -29,11 +30,14 @@
 import java.nio.channels.ClosedChannelException;
 import java.util.List;
 import java.util.Map;
+import org.apache.commons.lang.StringUtils;
 
 /** Common functionality shared by concrete classes. */
 abstract class AbstractHttpHandler<T extends HttpObject> extends SimpleChannelInboundHandler<T>
     implements ChannelOutboundHandler {
 
+  private static String USER_AGENT_VALUE = null;
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: why not keep the assignment of `BlazeVersionInfo.instance().getVersion()`?

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java [hunk 3]
```diff
@@ -78,6 +82,14 @@ protected void addCredentialHeaders(HttpRequest request, URI uri) throws IOExcep
     }
   }
 
+  protected void addUserAgentHeader(HttpRequest request) {
+    if (USER_AGENT_VALUE == null) {
+      USER_AGENT_VALUE =
+          "bazel/" + StringUtils.defaultIfEmpty(BlazeVersionInfo.instance().getVersion(), "dev");
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: as discussed yesterday I believe we should try to be consistent between the gRPC protocol and the http protocol. So I'd argue if no version is provided just have plain `bazel` and if a version is provided have `bazel <version number>`.

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java [hunk 4]
```diff
@@ -29,11 +30,14 @@
 import java.nio.channels.ClosedChannelException;
 import java.util.List;
 import java.util.Map;
+import org.apache.commons.lang.StringUtils;
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: We are trying to get rid of apache commons as a dependency in Bazel. It seems overkill to add a dependency to apache commons as a one liner.

File: src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java [hunk 5]
```diff
@@ -34,22 +33,22 @@
 
   @Test
   public void testEmptyVersionInfoMeansNotAvailable() {
-    BlazeVersionInfo info = new BlazeVersionInfo(Collections.<String, String>emptyMap());
+    BlazeVersionInfo info = new BlazeVersionInfo(ImmutableMap.of());
```
Comment (buchgr) [inline src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java]: it's typically good practice to not change unrelated code, as to keep a change as focused as possible. Feel free to send a separate clean up CL if you think this adds value :-).

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java (hunk 3 above)
Comment (buchgr) [reply src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: Also I don't think we use the `bazel/` notation anywhere else but typically just say `bazel <version>`. Is there a reason why we should use `/` in user agent?

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java (hunk 4 above)
Comment (ishikhman) [reply src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: Alright, noted :+1:

File: src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java (hunk 3 above)
Comment (ishikhman) [reply src/main/java/com/google/devtools/build/lib/remote/blobstore/http/AbstractHttpHandler.java]: As discussed, this is a standard format, see: https://tools.ietf.org/html/rfc7231#section-5.5.3 and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent

File: src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java (hunk 5 above)
Comment (ishikhman) [reply src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java]: agree :)

File: src/main/java/com/google/devtools/build/lib/analysis/BlazeVersionInfo.java [hunk 6]
```diff
@@ -117,12 +119,12 @@ public String getReleaseName() {
...
 
   /**
-   * Returns the version, if any, or HEAD. The returned version number is easier to
+   * Returns the version, if any, or {@code ""}. The returned version number is easier to
    * process than the version returned by #getReleaseName().
    */
   public String getVersion() {
     String buildLabel = buildData.get(BUILD_LABEL);
-    return buildLabel != null ? buildLabel : "";
+    return nullToEmpty(buildLabel);
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/analysis/BlazeVersionInfo.java]: nit: please undo the change here too then.

File: src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java [hunk 7]
```diff
@@ -34,22 +33,22 @@
...
-    Map<String, String> data = singletonMap("Build label", "");
+    Map<String, String> data = ImmutableMap.of("Build label", "");
     BlazeVersionInfo info = new BlazeVersionInfo(data);
     assertThat(info.getReleaseName()).isEqualTo("development version");
   }
 
   @Test
   public void testReleaseNameIfBuildLabelIsPresent() {
-    Map<String, String> data = singletonMap("Build label", "3/4/2009 (gold)");
+    Map<String, String> data = ImmutableMap.of("Build label", "3/4/2009 (gold)");
```
Comment (buchgr) [inline src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java]: same here

File: src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java [hunk 8]
```diff
@@ -34,22 +33,22 @@
...
+    BlazeVersionInfo info = new BlazeVersionInfo(ImmutableMap.of());
     assertThat(info.isAvailable()).isFalse();
     assertThat(info.getSummary()).isNull();
     assertThat(info.getReleaseName()).isEqualTo("development version");
   }
 
   @Test
   public void testReleaseNameIsDevelopmentIfBuildLabelIsNull() {
-    Map<String, String> data = singletonMap("Build label", "");
+    Map<String, String> data = ImmutableMap.of("Build label", "");
```
Comment (buchgr) [inline src/test/java/com/google/devtools/build/lib/runtime/BlazeVersionInfoTest.java]: same here
